securesystemslib/hsm.py

```diff
@@ -0,0 +1,473 @@
+HSM_SUPPORT = True
+NO_HSM_MSG = "HSM support requires PyKCS11 library"
+try:
+  import PyKCS11
+
+  # RSA-PSS with SHA256 hash to be used for signature generation.
+  RSA_PSS_MECH = PyKCS11.CKM_SHA256_RSA_PKCS_PSS
+  # SHA256 hash to be used to digest the data.
+  RSA_PSS_HASH_SHA256 = PyKCS11.CKM_SHA256
+  # Mask generating function for SHA256 Hash.
+  RSA_PSS_MGF_SHA256 = PyKCS11.CKG_MGF1_SHA256
+  # Length of salt to be used for hashing.
+  RSA_PSS_SALT_LENGTH = 32
+except ImportError:
+  HSM_SUPPORT = False
+
+logger = logging.getLogger(__name__)
+
+class HSM(object):
+  """
+  <Purpose>
+    Provides an interface to use cryptographic tokens for various
+    cryptographic operations.
+
+  <Arguments>
+    PKCS11Lib_path:
+       path to the PKCS#11 library. This can be module specific or
+       library by OpenSC(opensc-pkcs11.so) can be used.
+  <Exceptions>
+    securesystemslib.exceptions.NotFoundError, if the path of PKCS#11
+    library is not specified or the library is corrupt
+
+    securesystemslib.exceptions.UnsupportedLibraryError, if the PyKCS11 module
+    is not available.
+  """
+
+  def __init__(self, PKCS11Lib_path):
+
+    if not HSM_SUPPORT: # pragma: no cover
+      raise securesystemslib.exceptions.UnsupportedLibraryError(NO_HSM_MSG)
+
+    self.PKCS11LIB = PKCS11Lib_path
+
+    # Initialize the PyKCS11Lib, wrapper of PKCS#11 in Python.
+    self.PKCS11 = PyKCS11.PyKCS11Lib()
+
+    self.session = None
+
+    # Load the PKCS11 shared library file.
+    self.refresh()
+
+
+
+
+
+  def refresh(self):
+    """
+    <Purpose>
+      This method refreshes the list of available cryptographic tokens.
+
+    <Exceptions>
+      securesystemslib.exceptions.NotFoundError, if the path of PKCS#11
+      library is not specified or the library is corrupt
+    """
+
+    # Try to load the PKCS11 library
+    try:
+      # Load the PKCS#11 library and simultaneously update the list
+      # of available HSM.
+      self.PKCS11.load(self.PKCS11LIB)
+    except PyKCS11.PyKCS11Error as error:
+      logger.error('PKS11 Library not found or is corrupt!')
+      raise securesystemslib.exceptions.NotFoundError(error.__str__())
+
+
+
+
+
+  def get_available_HSMs(self):
+    """
+    <Purpose>
+      Generate the list of available cryptographic tokens for the user
+
+    <Exceptions>
+      securesystemslib.exceptions.NotFoundError, if the path of PKCS#11
+      library is not specified or the library is corrupt
+
+    <Returns>
+      A list of dictionaries consisting of relevant information
+      regarding all the available tokens
+    """
+
+    # Refresh the list of available slots for HSM
+    self.refresh()
+
+    # Get the list of slots on which HSMs are available
+    slot_list = self.PKCS11.getSlotList()
+    slot_info_list = []
+
+    # For all the available HSMs available, add relevant information
+    # to the slots dictionary
+    for slot in slot_list:
+      slot_dict = dict()
```

Comment:
More of a style nit, but literal construction is a little easier to read and less repetitive:

```python
slot_dict = {
    'slot_id': slot,
    'slot_info': self.PKCS11.getSlotInfo(slot),
    # ...
}
slot_info_list.append(slot_dict)
```